README.md
# NAME

Mojolicious::Plugin::DbicSchemaViewer - Viewer for DBIx::Class schema definitions

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.20+-blue.svg" alt="Requires Perl 5.20+" />
    <a href="https://travis-ci.org/Csson/p5-Mojolicious-Plugin-DbicSchemaViewer"><img src="https://api.travis-ci.org/Csson/p5-Mojolicious-Plugin-DbicSchemaViewer.svg?branch=master" alt="Travis status" /></a>
    <a href="http://cpants.cpanauthors.org/release/CSSON/Mojolicious-Plugin-DbicSchemaViewer-0.0103"><img src="http://badgedepot.code301.com/badge/kwalitee/CSSON/Mojolicious-Plugin-DbicSchemaViewer/0.0103" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Mojolicious-Plugin-DbicSchemaViewer%200.0103"><img src="http://badgedepot.code301.com/badge/cpantesters/Mojolicious-Plugin-DbicSchemaViewer/0.0103" alt="CPAN Testers result" /></a>
    <img src="https://img.shields.io/badge/coverage-59.0%-red.svg" alt="coverage 59.0%" />
    </p>
</div>

# VERSION

Version 0.0103, released 2016-09-20.

# SYNOPSIS

    $self->plugin(DbicSchemaViewer => {
        schema => Your::Schema->connect(...),
    });

# DESCRIPTION

This plugin is a definition viewer for [DBIx::Class](https://metacpan.org/pod/DBIx::Class) schemas. It currently offers two different views on the schema:

- It lists all result sources with column definitions and and their relationships in table form.
- It uses  [DBIx::Class::Visualizer](https://metacpan.org/pod/DBIx::Class::Visualizer) to generate an entity-relationship model.

## Configuration

The following settings are available. It is recommended to use either ["router"](#router) or ["condition"](#condition) to place the viewer behind some kind of authorization check.

### allowed\_schemas

An optional array reference consisting of schema classes. If set, only these classes are available for viewing.

If not set, all findable schema classes can be viewed.

### url

Optional.

By default, the viewer is located at `/dbic-schema-viewer`.

    $self->plugin(DbicSchemaViewer => {
        url => '/the-schema',
        schema => Your::Schema->connect(...),
    });

The viewer is instead located at `/the-schema`.

Note that the CSS and Javascript files are served under `/dbic-schema-viewer` regardless of this setting.

### router

Optional. Can not be used together with ["condition"](#condition).

Use this when you which to place the viewer behind an `under` route:

    my $secure = $app->routes->under('/secure' => sub {
        my $c = shift;
        return defined $c->session('logged_in') ? 1 : 0;
    });

    $self->plugin(DbicSchemaViewer => {
        router => $secure,
        schema => Your::Schema->connect(...),
    });

Now the viewer is located at `/secure/dbic-schema-viewer` (if the check is successful).

### condition

Optional. Can not be used together with ["router"](#router).

Use this when you have a named condition you which to place the viewer behind:

    $self->routes->add_condition(random => sub { return !int rand 4 });

    $self->plugin(DbicSchemaViewer => {
        condition => 'random',
        schema => Your::Schema->connect(...),
    });

# DEMO

There is a demo available at [http://dsv.code301.com/MadeUp::Book::Schema](http://dsv.code301.com/MadeUp::Book::Schema). Don't miss the help page for instructions.

# SEE ALSO

- `dbic-schema-viewer` - a small application (in `/bin`) for running this plugin standalone.
- [DBIx::Class::Visualizer](https://metacpan.org/pod/DBIx::Class::Visualizer)

# SOURCE

[https://github.com/Csson/p5-Mojolicious-Plugin-DbicSchemaViewer](https://github.com/Csson/p5-Mojolicious-Plugin-DbicSchemaViewer)

# HOMEPAGE

[https://metacpan.org/release/Mojolicious-Plugin-DbicSchemaViewer](https://metacpan.org/release/Mojolicious-Plugin-DbicSchemaViewer)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Mojolicious-Plugin-DbicSchemaViewer

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0200 2016-09-20T15:42:06Z UTC
 [Documentation]
 - The included static example file has been replaced with an online demo.

 [New Features]
 - The visualizer tab reimplemented to use improved
   DBIx::Class::Visualizer.

 [Requirements]
 - DBIx::Class::Visualizer is now required.

 [Stats & Dependency Changes]
 - 18 files changed, 953 insertions(+), 449 deletions(-)
 - (run req) + DBIx::Class::Visualizer 0.0200
 - (run req) + File::HomeDir (any)
 - (run req) + File::ShareDir 1.102
 - (run req) + PerlX::Maybe 1.001
 - (run req) + String::Random (any)
 - (run req) + String::Stomp 0.0102
 - (run req) + Syntax::Feature::Qi (any)
 - (run req) - File::ShareDir::Tarball
 - (dev req) + Dist::Zilla::Plugin::ShareDir (any)
 - (dev req) ~ Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0100
   --> 0.0200
 - (dev req) - Dist::Zilla::Plugin::ShareDir::Tarball

0.0102 2016-09-04T20:50:39Z UTC
 [Documentation]
 - Fixed some typos.

 [Enhancements]
 - If DBIx::Class::Visualized is installed, a graphical representation of
   the schema can be rendered.

 [Stats & Dependency Changes]
 - 16 files changed, 445 insertions(+), 68 deletions(-)
 - (run req) + DateTime::Tiny 1.06
 - (run req) + Mojolicious::Plugin::BootstrapHelpers 0.0201
 - (run req) ~ Path::Tiny (any) --> 0.072
 - (run req) ~ Safe::Isa (any) --> 1.000005
 - (test req) + IO::Handle (any)
 - (test req) + IPC::Open3 (any)
 - (conf req) + ExtUtils::MakeMaker (any)
 - (conf req) + File::ShareDir::Install 0.06
 - (conf req) - Module::Build::Tiny
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Badge::Depot::Plugin::Travis (any)
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.007
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0100
 - (dev req) + Dist::Zilla::Plugin::Clean 0.07
 - (dev req) + Dist::Zilla::Plugin::DistIller::MetaGeneratedBy (any)
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::Git::Contributors 0.017
 - (dev req) + Dist::Zilla::Plugin::InstallRelease 0.008
 - (dev req) + Dist::Zilla::Plugin::MakeMaker (any)
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale (any)
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Dist::Zilla::Plugin::Test::Compile 2.054
 - (dev req) + Dist::Zilla::Plugin::Test::EOF 0.0501
 - (dev req) + Dist::Zilla::Plugin::Test::EOL 0.18
 - (dev req) + Dist::Zilla::Plugin::Test::Kwalitee::Extra v0.2.1
 - (dev req) + Dist::Zilla::Plugin::Test::NoTabs 0.15
 - (dev req) + Dist::Zilla::Plugin::Test::Version 1.05
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::EOL (any)
 - (dev req) + Test::More 0.88
 - (dev req) + Test::NoTabs 1.4
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.151680
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.003001
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.150250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.028
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs (any) --> 0.021
 - (dev req) ~ Dist::Zilla::Plugin::TravisYML (any) --> 1.15
 - (dev req) ~ Pod::Weaver::Section::Badges (any) --> 0.0401
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Zilla::Plugin::CheckExtraTests
 - (dev req) - Dist::Zilla::Plugin::ContributorsFromGit
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::ModuleBuildTiny
 - (dev req) - Dist::Zilla::Plugin::NextRelease
 - (dev req) - Dist::Zilla::Plugin::NextVersion::Semantic
 - (dev req) - Dist::Zilla::Plugin::OurPkgVersion
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Dist::Zilla::Plugin::PreviousVersion::Changelog
 - (dev req) - Dist::Zilla::Plugin::ReversionOnRelease
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) + Dist::Iller 0.1408
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0312

 [Testing]
 - Expanded the test schema (which is also used as an example).

0.0101 2015-10-28T19:08:49Z UTC
 [Documentation]
 - Fixed bad internal links.

0.0100 2015-10-28T18:21:23Z UTC
 - Initial release


t/lib/TestFor/MPDbicSchemaViewer/Schema.pm
package TestFor::MPDbicSchemaViewer::Schema;

use base 'DBIx::Class::Schema';

sub schema_version { 2 }

__PACKAGE__->load_namespaces;

1;


t/lib/TestFor/MPDbicSchemaViewer/Schema/Result/One.pm
package TestFor::MPDbicSchemaViewer::Schema::Result::One;

use base 'DBIx::Class::Core';

__PACKAGE__->table('One');
__PACKAGE__->add_columns(
    one_id => { data_type => 'int' },
    thedata => { data_type => 'varchar' },
);

__PACKAGE__->set_primary_key(qw/one_id/);

__PACKAGE__->has_many(many => 'TestFor::MPDbicSchemaViewer::Schema::Result::Many', 'one_id');

1;


t/lib/TestFor/MPDbicSchemaViewer/Schema/Result/Book.pm
package TestFor::MPDbicSchemaViewer::Schema::Result::Book;

use base 'DBIx::Class::Core';

__PACKAGE__->table('Book');
__PACKAGE__->add_columns(
    book_id => {
        data_type => 'int',
        is_auto_increment => 1,
    },
    isbn => {
        data_type => 'varchar',
        size => 13,
    },
    title => {
        data_type => 'varchar',
    },
);

__PACKAGE__->set_primary_key(qw/book_id/);

__PACKAGE__->has_many(book_authors => 'TestFor::MPDbicSchemaViewer::Schema::Result::BookAuthor', 'book_id');
__PACKAGE__->many_to_many(authors => 'book_authors', 'author_id');

1;


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.031.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Viewer for DBIx::Class schemas",
  "dist_author" => [
    "Erik Carlsson <info\@code301.com>"
  ],
  "dist_name" => "Mojolicious-Plugin-DbicSchemaViewer",
  "dist_version" => "0.0001",
  "license" => "perl",
  "module_name" => "Mojolicious::Plugin::DbicSchemaViewer",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "perl" => "5.020000"
  },
  "script_files" => [],
  "share_dir" => {
    "dist" => "templates"
  },
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.96"
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Build" => "0.3601",
  "Test::More" => "0.96"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


share/public/dbic-schema-viewer/svg.js
var pressedKeys = {
    ctrl: false,
    shift: false,
    alt: false,
};
$(document).ready(function() {
    var $doc = $('svg');

    $doc.find('.edge')
        .hover(function(e) {
            var $edge = $(this);

            $edge.addClass('hovered');
            $edge.appendTo($edge.parent());
            $edge.css('cursor', 'pointer');
        })
        .mouseout(function(e) {
            var $edge = $(this);

            if(!$edge.attr('data-clicked')) {
                $edge.removeClass('hovered');
            }
        })
        .click(function(e) {
            var $edge = $(this);
            if($edge.attr('data-clicked')) {
                $edge.removeAttr('data-clicked');
            }
            else {
                $edge.attr('data-clicked', 1);
            }
        });

    // Handle column name hovers
    $doc.find('polygon.column-name, text.column-name')
        .mouseenter(function(e) {
            var $selected = $(this);
            var $text = $selected.prop('tagName') === 'text' ? $selected : $selected.siblings(byColumnTextIdFromPolygon($selected));
            var $polygon = $selected.prop('tagName') === 'polygon' ? $selected : $selected.siblings(byColumnPolygonIdFromText($selected));

            $polygon.addClass('hovered');
            highlightRelation($doc, $text.parent().data('table-name'), $text.data('column-name'));
        })
        .mouseout(function(e) {
            var $selected = $(this);
            var $text = $selected.prop('tagName') === 'text' ? $selected : $selected.siblings(byColumnTextIdFromPolygon($selected));
            var $polygon = $selected.prop('tagName') === 'polygon' ? $selected : $selected.siblings(byColumnPolygonIdFromText($selected));

            if($text.attr('data-clicked')) {
                return;
            }
            $polygon.removeClass('hovered');
            removeDependentHighlights($doc, $text.parent().data('table-name') + '.' + $text.data('column-name'));
        })
        .click(function(e) {
            var $text = $(this).prop('tagName') === 'text' ? $(this) : $(this).siblings(byColumnTextIdFromPolygon($(this)));

            if($text.attr('data-clicked')) {
                $text.removeAttr('data-clicked');
            }
            else {
                $text.attr('data-clicked', 1);
            }
        });

    // Highlight all relations to/from a table when the table name cell is hovered.
    $doc.find('polygon.table-name, text.table-name')
        .mouseenter(function(e) {
            var $node = $(this).parent();

            if(!$node.hasClass('unwanted')) {
                $node.addClass('hovered').removeClass('faded');
                highlightRelation($doc, $node.data('table-name'));
            }
        })
        .mouseout(function(e) {
            var $node = $(this).parent();

            if($node.hasClass('unwanted')) {
                $node.removeClass('hovered');
            }
            if($node.attr('data-clicked')) {
                return;
            }
            $node.removeClass('hovered');
            removeDependentHighlights($doc, $node.data('table-name'));
        })
        .click(function(e) {
            var $node = $(this).parent();

            var params = allParams();
            if(pressedKeys['shift'] && pressedKeys['ctrl']) {
                var skip = params['skip_result_source_names'] ? params['skip_result_source_names'].split(',') : [];
                skip.push($node.data('table-name'));
                params['skip_result_source_names'] = skip.join(',');

                //params['skip_result_source_names'] = $node.data('table-name');
                window.location = paramToQueryString(params);
            }
            else if(pressedKeys['shift']) {
                params['wanted_result_source_names'] = $node.data('table-name');
                window.location = paramToQueryString(params);
            }

            if($node.attr('data-clicked')) {
                $node.removeAttr('data-clicked');
            }
            else {
                $node.attr('data-clicked', 1);
            }
            if(pressedKeys['ctrl'] && $node.attr('data-clicked')) {
                $node.addClass('unwanted').removeClass('hovered');
                removeDependentHighlights($doc, $node.data('table-name'));
            }
            else {
                    $node.removeClass('unwanted');
                    highlightRelation($doc, $node.data('table-name'));
            }
    });

    $doc.find('.node *')
        .mouseenter(function() {
            $(this).parent().addClass('mouseover').removeClass('faded');
        })
        .mouseout(function() {
            $(this).parent().removeClass('mouseover');
        });

    $(document).keyup(function(e) {
        // escape -> restore (but only if the #help-modal isn't opened)
        if(e.keyCode === 27 && $('#help-modal').css('display') === 'none') {
            $doc.find('.edge').removeClass('hovered active').removeAttr('data-clicked');
            $doc.find('.node').removeClass('hovered related-table faded unwanted').removeAttr('data-clicked');
            $doc.find('.node polygon').removeClass('hovered').removeAttr('data-clicked');
            $doc.find('text').removeClass('hovered').removeAttr('data-clicked');
            $doc.find('[data-highlighted-by]').removeAttr('data-highlighted-by');
        }
        // r -> zoom to selected tables
        else if(isKey(e, 'R')) {
            var params = allParams();
            var wanted = [];
            var skip = params['skip_result_source_names'] ? params['skip_result_source_names'].split(',') : [];

            $doc.find('.node.hovered').each(function() {
                wanted.push($(this).data('table-name'));
            });
            $doc.find('.node.unwanted').each(function() {
                skip.push($(this).data('table-name'));
            })

            if(wanted.length) {
                params['wanted_result_source_names'] = wanted.join(',');
            }
            if(skip.length) {
                params['skip_result_source_names'] = skip.join(',');
            }
            window.location = paramToQueryString(params);
        }
        // q -> back to overall visualization
        else if(isKey(e, 'Q')) {
            window.location = '?schema=' + param('schema');
        }
        // 0..9 -> degrees_of_separation
        else if(e.keyCode >= 48 && e.keyCode <= 57) {
            e.preventDefault();
            var params = allParams();
            var degrees = e.keyCode - 48;

            if(pressedKeys['shift']) {
                degrees += 10;
            }
            if(pressedKeys['ctrl']) {
                degrees += 20;
            }
            params['degrees_of_separation'] = degrees;
            window.location = paramToQueryString(params);
        }
    });

    // remember shift/ctrl status
    $(document).keydown(function(e) {
            if(e.ctrlKey) {
                pressedKeys['ctrl'] = true;
            }
            if(e.shiftKey) {
                pressedKeys['shift'] = true;
            }
        })
        .keyup(function(e) {
            if(!e.ctrlKey) {
                pressedKeys['ctrl'] = false;
            }
            if(!e.shiftKey) {
                pressedKeys['shift'] = false;
            }
        });

    if(param('wanted_result_source_names')) {
        var tableNames = param('wanted_result_source_names').split(',');
        var findString = tableNames.map(tableName => '.node' + byAttr('data-table-name', tableName) + ' polygon.table-name').join(', ');
        $(findString).mouseenter().click();
    }

});
// call byAttr('data-column-name', $el)     instead of  '[data-column-name="' + $el.attr('data-column-name') + '"]'
//   or byAttr('data-column-name', string)  instead of  '[data-column-name="' + string + '"]'
function byAttr(attr, thing, operand) {
    var string = typeof thing === 'object' ? thing.attr(attr) : thing;
    var operand = operand || '=';
    return '[' + attr + operand + '"' + string + '"]';
}

// polygon.column-name has id 'bg-column-%s'
// text.column-name    has id 'column-%s'
function byColumnTextIdFromPolygon($polygon) {
    return '#' + $polygon.attr('id').replace(/^bg-/, '');
}
function byColumnPolygonIdFromText($text) {
    return '#bg-' + $text.attr('id');
}

function addHighlightedBy($el, by) {
    if($el.attr('data-highlighted-by')) {
        var currentvalue = $el.attr('data-highlighted-by');
        if(!currentvalue.match(new RegExp('#' + by + '#'))) {
            var newValue = $el.attr('data-highlighted-by') + '#' + by + '#';
            $el.attr('data-highlighted-by', newValue);
        }
    }
    else {
        $el.attr('data-highlighted-by', '#' + by + '#');
    }
}
function removeDependentHighlights($doc, by) {
    by = '#' + by + '#';
    $doc.find('.edge' + byAttr('data-highlighted-by', by, '*=')).each(function() {
        var $edge = $(this);
        _removeHighlightHelper($edge, by);
        if(!$edge.attr('data-highlighted-by')) { $edge.removeClass('active'); }
    });
    $doc.find('.node' + byAttr('data-highlighted-by', by, '*=')).each(function() {
        var $node = $(this);
        _removeHighlightHelper($node, by);
        if(!$node.attr('data-highlighted-by')) { $node.removeClass('related-table'); }
    });

    // Add fadedness to nodes if any other node is a .related-table.
    // Otherwise, remove fadedness from all.
    if($doc.find('.node.related-table').length) {
        $doc.find('.node:not(.related-table):not(.hovered):not(.unwanted)').addClass('faded');
    }
    else {
        $doc.find('.node.faded').removeClass('faded');
    }
}
// by must be '#by#'
function _removeHighlightHelper($el, by) {
    $el.attr('data-highlighted-by', $el.attr('data-highlighted-by').replace(new RegExp(by, 'g'), ''));
    if(!$el.attr('data-highlighted-by')) {
        $el.removeAttr('data-highlighted-by');
    }
}
function getEdgesByTableColumn(table, column) {
    return '.edge' + byAttr('data-origin-table', table) + byAttr('data-origin-column', column)
       + ', .edge' + byAttr('data-destination-table', table) + byAttr('data-destination-column', column);
}
function highlightRelation($doc, tableName, columnName) {
    var byOrigin = byAttr('data-origin-table', tableName);
    var byDestination = byAttr('data-destination-table', tableName);

    if(undefined !== columnName) {
        byOrigin = byOrigin + byAttr('data-origin-column', columnName);
        byDestination = byDestination + byAttr('data-destination-column', columnName);
    }
    var by = tableName + (columnName !== undefined ? '.'+columnName : '');
    $doc.find('.edge' + byOrigin + ', .edge' + byDestination).each(function(index) {
        var $edge = $(this);
        $edge.addClass('active');
        $edge.appendTo($edge.parent());

        var otherTableName = tableName === $edge.data('origin-table') ? $edge.data('destination-table') : $edge.data('origin-table');
        var $otherTable = $doc.find('.node' + byAttr('data-table-name', otherTableName));

        addHighlightedBy($edge, by);
        addHighlightedBy($otherTable, by);
        $otherTable.addClass('related-table').removeClass('faded');
    });

    if($doc.find('.node.related-table').size()) {
        $doc.find('.node:not(.related-table):not(.hovered):not(mouseover):not(.unwanted)').addClass('faded');
    }
}
function isKey(event, wantedKey) {
    return wantedKey.charCodeAt() === event.keyCode;
}
function allParams() {
    var queryString = {};
    location.search.substr(1)
                   .split('&')
                   .filter(string => string.length)
                   .map(qs => qs.split('='))
                   .map(listOfPairs => queryString[listOfPairs[0]] = listOfPairs[1]);
    return queryString;
}
function param(key) {
    var params = allParams();
    return params[key];
}
function paramToQueryString(params) {
    if(undefined === params) {
        params = allParams();
    }

    return '?' + Object.keys(params).map(key => { return [key, params[key]].join('=') }).join('&');
}

Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.007.
use strict;
use warnings;

use 5.020000;

use ExtUtils::MakeMaker;

use File::ShareDir::Install;
$File::ShareDir::Install::INCLUDE_DOTFILES = 1;
$File::ShareDir::Install::INCLUDE_DOTDIRS = 1;
install_share dist => "share";


my %WriteMakefileArgs = (
  "ABSTRACT" => "Viewer for DBIx::Class schema definitions",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::ShareDir::Install" => "0.06"
  },
  "DISTNAME" => "Mojolicious-Plugin-DbicSchemaViewer",
  "EXE_FILES" => [
    "bin/dbic-schema-viewer"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.020000",
  "NAME" => "Mojolicious::Plugin::DbicSchemaViewer",
  "PREREQ_PM" => {
    "DBIx::Class::Visualizer" => "0.0200",
    "Data::Dump::Streamer" => 0,
    "DateTime::Tiny" => "1.06",
    "File::HomeDir" => 0,
    "File::ShareDir" => "1.102",
    "Mojolicious" => "6.00",
    "Mojolicious::Plugin::BootstrapHelpers" => "0.0201",
    "Path::Tiny" => "0.072",
    "PerlX::Maybe" => "1.001",
    "Safe::Isa" => "1.000005",
    "String::Random" => 0,
    "String::Stomp" => "0.0102",
    "Syntax::Feature::Qi" => 0,
    "experimental" => 0
  },
  "TEST_REQUIRES" => {
    "DBIx::Class" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Deep" => 0,
    "Test::Mojo" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.0201",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DBIx::Class" => 0,
  "DBIx::Class::Visualizer" => "0.0200",
  "Data::Dump::Streamer" => 0,
  "DateTime::Tiny" => "1.06",
  "ExtUtils::MakeMaker" => 0,
  "File::HomeDir" => 0,
  "File::ShareDir" => "1.102",
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Mojolicious" => "6.00",
  "Mojolicious::Plugin::BootstrapHelpers" => "0.0201",
  "Path::Tiny" => "0.072",
  "PerlX::Maybe" => "1.001",
  "Safe::Isa" => "1.000005",
  "String::Random" => 0,
  "String::Stomp" => "0.0102",
  "Syntax::Feature::Qi" => 0,
  "Test::Deep" => 0,
  "Test::Mojo" => 0,
  "Test::More" => "0.96",
  "experimental" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}
